Add case tile configuration (limited)

NoahCarnahan opened the pull request:
Allows case tiles to be used in the case list instead of the standard table layout. This feature is under a flag. It is also a "minimal" implementation intended to support TDH's use case only. Users map case properties to tile fields with the interface pictured below. The display text of the case property is used as a prefix to the value on the tile. Users also choose if the case tile will continue to show while filling out forms.

Configuration limitations:
- 5 fields on the case tile can be assigned to.
- The `sex` field must be assigned to an ID mapping with keys "male" and "female".
- The `date` filed must be a date. It will be converted to a string like "4 jours," "5 moins," or "2 ans" representing the difference between the current date and the property's value
- There are also images and actions in the case tile that are not configurable

The configuration interface:
![screen shot 2015-02-11 at 1 04 33 pm](https://cloud.githubusercontent.com/assets/2117127/6146028/0139f406-b1f1-11e4-83cf-3d2694931eb6.png)
The tile:
![screenshot_2015-02-10-08-57-56](https://cloud.githubusercontent.com/assets/2117127/6146049/3e306944-b1f1-11e4-842d-8600a6c9e432.png)

NoahCarnahan committed:
[WIP] Beginings of case tile template feature

- Added feature flag
- Added don't use/do use case tiles select
- Added case tile bucket selects to case properties list

NoahCarnahan committed:
[WIP] Save tile configuration on server

NoahCarnahan committed:
[WIP] enable save button when case tile configuration changed

NoahCarnahan committed:
Populate case tile template

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

Conflicts:
	corehq/apps/app_manager/templates/app_manager/partials/case_list_properties.html

NoahCarnahan committed:
Fix bug when flag is disabled and rename case_name bucket to child_name

NoahCarnahan committed:
Fix bug where save button was activated when nothing had changed

NoahCarnahan committed:
Fix issue where "name" was not converted to "case_name"

NoahCarnahan committed:
Add 'persist tile on forms' checkbox and save configuration in db

NoahCarnahan committed:
[WIP] add detail persistence to case xml (doesn't currently work)

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

NoahCarnahan committed:
update path checking

NoahCarnahan committed:
Merge branch 'fix-build-import' into tdh-tiles

NoahCarnahan committed:
Take tile field prefixes from headers. Organize imports

NoahCarnahan committed:
Revert "Merge branch 'fix-build-import' into tdh-tiles"

This reverts commit 0aad836051448be68522798cbe5e1608be1e4ea0, reversing
changes made to efa3a705ed743b29fa2b6000701b6ccc93657ff2.

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

NoahCarnahan committed:
Revert "Revert "Merge branch 'fix-build-import' into tdh-tiles""

This reverts commit d5670212776763480c4aeab1f79eade02d354405.

NoahCarnahan committed:
Revert "update path checking"

This reverts commit 1ff53150c842be0de131a28b41d4edf09055dcc0.

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

biyeun committed:
merge with master and resolve conflict with case-details-screen.js

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

Conflicts:
	corehq/apps/app_manager/static/app_manager/js/detail-screen-config.js

NoahCarnahan committed:
Give generic names to case tile buckets

NoahCarnahan committed:
Merge remote-tracking branch 'origin' into tdh-tiles

NoahCarnahan committed:
Merge remote-tracking branch 'origin/tdh-tiles' into tdh-tiles

NoahCarnahan committed:
Rename some of the tile fields again

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,60 @@ def details(self):
+                                        template_args[template_field]["enum_keys"] = {}
+                                        for mapping in column.enum:
+                                            key = mapping.key
+                                            template_args[template_field]["enum_keys"][key] = self.id_strings.detail_column_enum_variable(module, detail_type, column, key)
```
E501 line too long (171 > 115 characters)

dimagimon commented on corehq/apps/app_manager/static/app_manager/js/detail-screen-config.js:
```diff
@@ -300,6 +301,9 @@ var DetailScreenConfig = (function () {
             ]).val(this.original.model);
+
+            var icon = (CC_DETAIL_SCREEN.isAttachmentProperty(this.original.field)
+               ? COMMCAREHQ.icons.PAPERCLIP : null);
```
Bad line breaking before '?'.

dimagimon replied:
Bad line breaking before '?'.

NoahCarnahan committed:
PEP8 compliance

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+                                            template_args[template_field]["enum_keys"][key] = \
+                                                self.id_strings.detail_column_enum_variable(
+                                                    module, detail_type, column, key
+                                                )
```
E123 closing bracket does not match indentation of opening bracket's line

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+                                )
                                 r.append(d)
+
+                            elif detail.use_case_tiles:
```
It would be nice to refactor this block into a separate function. As is it's very deeply nested and makes the function pretty long, which makes a pretty bad smell.

millerdev replied:
Had to stop reviewing here (dinner time). Will continue later.

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+                                            column.header.get(self.app.default_language, "")
+                                        )
+                                    }
+                                    if column.format == "enum":
```
Looks like `column` could be `None` here. What do you think about writing some tests to exercise this code?

NoahCarnahan replied:
Added build validation and a test to confirm it catches this.

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+
+                                # Populate the template
+                                with open(os.path.join(
+                                    os.path.dirname(__file__), "case_tile_templates", "tdh.txt"
```
Would be better to read this file once outside the double nested `for` loops. But more generally, is this the best practice for opening a template? This also looks like code for a specific customer, which typically does not belong in the `corehq` package.

NoahCarnahan replied:
Update ensures file is read only once.

NoahCarnahan replied:
I'm not really sure what the best practices are for opening a template like this. Do you have any ideas?

The question of `corehq` is also an interesting one. I haven't done any work that hasn't gone into `corehq` before, so I don't know what that process generally is for custom work. After glancing at the `custom` directory, it looks like most of the things there are models or reports. I don't know if changes to the case list page and to `suite_xml.py` are as easy to pull out of the core. Thoughts?

millerdev replied:
@czue is there a precedent for customer-specific extensions to `corehq`?

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+                                    for col in detail.columns:
+                                        if col.case_tile_field == template_field:
+                                            column = col
+                                            break
```
It would be more efficient these detail columns once making a dict, and then use that to look up each template field:
```python
cols_by_tile = {col.case_tile_field: col for col in detail.columns}
for template_field in ...
    column = cols_by_tyle.get(template_field)
```

NoahCarnahan replied:
Good point, thanks! Updated in latest commits.

millerdev commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -920,8 +926,63 @@ def details(self):
+                                    template = f.read()
+                                detail_as_string = template.format(**template_args)
+
+                                detail = load_xmlobject_from_string(
```
Is it intentional that you're reusing a variable name here?

NoahCarnahan replied:
Fixed.

millerdev commented on corehq/apps/app_manager/views.py:
```diff
@@ -1536,8 +1538,12 @@ def edit_module_detail_screens(req, domain, app_id, module_id):
+    if use_case_tiles is not None and short is not None:
+        detail.short.use_case_tiles = use_case_tiles
+    if persist_tile_on_forms is not None and short is not None:
+        detail.short.persist_tile_on_forms = persist_tile_on_forms
```
Looks like these two if blocks could be moved up under line 1530 `if short is not None:`, and then they would not need to repetitively check that condition.

NoahCarnahan replied:
Done, thanks.

NoahCarnahan committed:
Reorganize additions to suite_xml.py

- Move the case tile detail building code into its own method
- Save case tile template string so that the file is opened only once
- Raise an exception if a DetailColumn can't be found for each tile field

NoahCarnahan committed:
Add some case tile configuration validation to module validation.

This could be more comprehensive, but holding off for now given the
limited scope of this feature for now.

NoahCarnahan committed:
Remove repeated condition checks in edit_module_detail_screens

NoahCarnahan committed:
Add tests

- Confirm that detail is produced correctly with a basic configuration
- Confirm that App.validate_app catches missing fields in case tile configuration

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1020,6 +1038,70 @@ def due_date(due_days):
+                    template_args[template_field]["enum_keys"][key] = \
+                        self.id_strings.detail_column_enum_variable(
+                            module, detail_type, column, key
+                        )
```
E123 closing bracket does not match indentation of opening bracket's line

czue commented:
@millerdev not sure where your comment went. we've done it before. definitely not something we like to do but there are times when it's the least bad way to solve a particular problem for a project.

millerdev commented:
@czue we were discussing the template reading code at [suite_xml.py:1101](https://github.com/dimagi/commcare-hq/pull/5507/files#diff-27910c0ed3e4d6d075211ffc4db6cab4R1101). The rest of the comments are under the second oudated diff on this PR.
